src/snowflake/connector/s3_storage_client.py

```diff
@@ -197,48 +263,83 @@ def _send_request_with_authentication_and_retry(
             x_amz_headers = {}
-        if not headers:
+        if headers is None:
             headers = {}
+        if payload is None:
+            payload = b""
+        if query_parts is None:
+            query_parts = {}
+        parsed_url = urlparse(url)
+        x_amz_headers["x-amz-security-token"] = self.credentials.creds.get(
+            "AWS_TOKEN", ""
+        )
+        x_amz_headers["host"] = parsed_url.hostname
+        if unsigned_payload:
+            x_amz_headers["x-amz-content-sha256"] = UNSIGNED_PAYLOAD
+        else:
+            x_amz_headers["x-amz-content-sha256"] = (
+                SnowflakeS3RestClient._hash_bytes_hex(payload).lower().decode()
+            )
 
-        def generate_authenticated_url_and_args() -> Tuple[bytes, Dict[str, bytes]]:
+        def generate_authenticated_url_and_args_v4() -> Tuple[bytes, Dict[str, bytes]]:
             t = datetime.utcnow()
             amzdate = t.strftime("%Y%m%dT%H%M%SZ")
-
-            if "AWS_TOKEN" in self.credentials.creds:
-                x_amz_headers["x-amz-security-token"] = self.credentials.creds.get(
-                    "AWS_TOKEN"
-                )
-            _x_amz_headers = self.construct_canonicalized_headers(x_amz_headers)
-            string_to_sign = self._construct_string_to_sign(
-                verb, resources, _x_amz_headers, amzdate, content_type=content_type
+            short_amzdate = amzdate[:8]
+            x_amz_headers["x-amz-date"] = amzdate
+
+            (
+                canonical_request,
+                signed_headers,
+            ) = self._construct_canonical_request_and_signed_headers(
+                verb=verb,
+                canonical_uri_parameter=parsed_url.path
+                + (f";{parsed_url.params}" if parsed_url.params else ""),
+                query_parts=query_parts,
+                canonical_headers=x_amz_headers,
+                payload_hash=x_amz_headers["x-amz-content-sha256"],
+            )
+            string_to_sign, scope = self._construct_string_to_sign(
+                self.region_name,
+                "s3",
+                amzdate,
+                short_amzdate,
+                self._hash_bytes_hex(canonical_request.encode("utf-8")).lower(),
             )
-            signature = self.sign(
-                self.credentials.creds["AWS_SECRET_KEY"].encode("UTF-8"), string_to_sign
+            kDate = self._sign_bytes(
+                ("AWS4" + self.credentials.creds["AWS_SECRET_KEY"]).encode("utf-8"),
+                short_amzdate,
             )
-            authorization_header = (  # TODO
-                "AWS " + self.credentials.creds["AWS_KEY_ID"] + ":" + signature.decode()
+            kRegion = self._sign_bytes(kDate, self.region_name)
+            kService = self._sign_bytes(kRegion, "s3")
+            signing_key = self._sign_bytes(kService, "aws4_request")
+
+            signature = self._sign_bytes_hex(signing_key, string_to_sign).lower()
```

Comment:
So we're chaining the signing key and generating it at each step. I am assuming this is Amazon API standard.

Reply:
Yup https://docs.aws.amazon.com/general/latest/gr/sigv4-calculate-signature.html

Reply:
> I thought everything was HTTPS

It is, this is an option where we can skip hashing the payload into the canonical request which then gets hashed into the string to sign and makes it to the authorization header.

We use this only during PUTs, so that we don't need to hash each file that we are putting. I checked and this is also what `boto3` does.

In detail what this does is it replaces the payload's hash in the canonical request with `UNSIGNED-PAYLOAD` which than cascades like normal.